// angular-calendar-scheduler/esm2015/modules/scheduler/calendar-scheduler-event-title.component.js
import { Component, Input } from '@angular/core';
export class CalendarSchedulerEventTitleComponent {
    constructor() {
        this.showStatus = true;
        this.showContent = true;
        this.showActions = true;
    }
}
CalendarSchedulerEventTitleComponent.decorators = [
    { type: Component, args: [{
                selector: 'calendar-scheduler-event-title',
                template: `
        <ng-template #defaultTemplate>
            <div
                class="cal-scheduler-event-title"
                [innerHTML]="event.title | schedulerEventTitle:view:event">
            </div>
            <calendar-scheduler-event-actions
                *ngIf="showActions && !showContent && (event.isClickable || event.isDisabled)"
                class="no-content-actions"
                [event]="event">
            </calendar-scheduler-event-actions>
            <div *ngIf="event.status && showStatus"
                class="cal-scheduler-event-status"
                [class.ok]="event.status === 'ok'"
                [class.warning]="event.status === 'warning'"
                [class.danger]="event.status === 'danger'">
            </div>
        </ng-template>
        <ng-template
            [ngTemplateOutlet]="customTemplate || defaultTemplate"
            [ngTemplateOutletContext]="{
                view: view,
                event: event,
                showStatus: showStatus,
                showContent: showContent,
                showActions: showActions
            }">
        </ng-template>
    `,
                host: {
                    'class': 'cal-scheduler-event-title-container'
                }
            },] }
];
CalendarSchedulerEventTitleComponent.propDecorators = {
    view: [{ type: Input }],
    event: [{ type: Input }],
    showStatus: [{ type: Input }],
    showContent: [{ type: Input }],
    showActions: [{ type: Input }],
    customTemplate: [{ type: Input }]
};
